### PR TITLE
New consent Google API (EXPOSUREAPP-4635)

### PR DESCRIPTION
**Pull request by mtwalli**
### Google play services version that is required by this PR: `18210214000`

- Trigger Google consent Dialog when needed to share keys 
- Fallback if unrecoverable error happened is to use older Api for getting the keys
- Old consent boolean flag in the App is not affected by this change regardless of user selection 😒 
- Ticket is addressing the consent only before scanning QR-Code, while I show when it is (needed and possible) in other places such as warn others.
- In case of disallowance, this can be revoked from Google settings -> switch on/off

### Note: 
`requestPreAuthorizedTemporaryExposureKeyHistory` and `requestPreAuthorizedTemporaryExposureKeyRelease`
are used always in combination. i.e:
1- `requestPreAuthorizedTemporaryExposureKeyRelease` can not succeed unless  `requestPreAuthorizedTemporaryExposureKeyHistory` is called before. 
2- `requestPreAuthorizedTemporaryExposureKeyHistory` is triggering the consent dialog once as long as `requestPreAuthorizedTemporaryExposureKeyRelease` was not called at least once in the next 5 days.

Testing: test the submission flow as before. This change require Google play services Update. Old versions should still work and use older Apis

### Important notes:
1- Resetting the App -> does not revoke the consent
2- Consent is completely independent from Test results (positive - pending - negative) -> therefore if you saw it once that is OK (you will be asked when ever this possible)

### TAN scenarios: 
1- if you were about to use QR code scan and you saw the consent dialog and took an action that is enough for any other case that requires consent such as TAN -> we can only show the consent once using the new Api
2- if you have not seen the consent before ,you should see it when following TAN process , later after you enter it and proceed to warn others

### Some scenarios: 

Case 1: Happy path user allows the consent and proceed to sharing the keys.
https://user-images.githubusercontent.com/25054729/107620898-b8993380-6c55-11eb-9447-b1927f3796a1.mp4

Case 2: User dis-allow form the consent  ,we proceed anyway to the keys sharing using the old api
https://user-images.githubusercontent.com/25054729/107625156-32342000-6c5c-11eb-88e6-e0dba01f40db.mp4

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/sharing/ExposureSharingService.kt [hunk 1]
```diff
@@ -1,73 +0,0 @@
...
-import com.google.zxing.BarcodeFormat
-import com.google.zxing.qrcode.QRCodeWriter
-import de.rki.coronawarnapp.exception.ExceptionCategory.EXPOSURENOTIFICATION
-import de.rki.coronawarnapp.exception.ExceptionCategory.INTERNAL
-import de.rki.coronawarnapp.exception.reporting.report
-import de.rki.coronawarnapp.server.protocols.AppleLegacyKeyExchange
-import de.rki.coronawarnapp.util.di.AppInjector
-import timber.log.Timber
-
-object ExposureSharingService {
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/sharing/ExposureSharingService.kt:17]: Was not used at all

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/consent/SubmissionConsentViewModel.kt [hunk 2]
```diff
@@ -24,7 +31,24 @@ class SubmissionConsentViewModel @AssistedInject constructor(
 
     fun onConsentButtonClick() {
         submissionRepository.giveConsentToSubmission()
-        routeToScreen.postValue(SubmissionNavigationEvents.NavigateToQRCodeScan)
+        launch {
+            try {
+                val preAuthorized = tekHistoryProvider.preAuthorizeExposureKeyHistory()
+                // Routes to QR code screen either has already granted permission or it is older Api
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/qrcode/consent/SubmissionConsentViewModel.kt:37]: `user has ...`

File: Corona-Warn-App/build.gradle [hunk 3]
```diff
@@ -385,7 +386,7 @@ dependencies {
     androidTestImplementation "androidx.navigation:navigation-testing:$nav_version"
 
     // Play Services
-    implementation 'com.google.android.play:core:1.9.0'
+    implementation 'com.google.android.play:core:1.9.1'
```
Comment (mtwalli) [inline Corona-Warn-App/build.gradle:389]: revert

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/version/DefaultENFVersion.kt [hunk 4]
```diff
@@ -41,8 +40,6 @@ class DefaultENFVersion @Inject constructor(
     }
 
     override suspend fun isAtLeast(compareVersion: Long): Boolean {
-        if (!compareVersion.isCorrectVersionLength) throw IllegalArgumentException("given version has incorrect length")
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/version/DefaultENFVersion.kt:44]: Comparing based on length is not valid anymore. latest version has longer length. I think Long comparison is enough here

**Comment by mtwalli** [comment]
Some scenarios: 


https://user-images.githubusercontent.com/25054729/107620898-b8993380-6c55-11eb-9447-b1927f3796a1.mp4


https://user-images.githubusercontent.com/25054729/107620573-34df4700-6c55-11eb-904f-da4934f30d4c.mp4




Uploading device-2021-02-11-102541.mp4…

File: Corona-Warn-App/src/main/AndroidManifest.xml [hunk 5]
```diff
@@ -64,6 +64,7 @@
         <activity
             android:name=".ui.main.MainActivity"
             android:exported="false"
+            android:launchMode="singleInstance"
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/AndroidManifest.xml:67]: New Exposure notification was triggering another instance of the MainAcitivty.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/tekhistory/DefaultTEKHistoryProvider.kt [hunk 6]
```diff
@@ -15,16 +26,53 @@ import kotlin.coroutines.suspendCoroutine
...
+    override suspend fun getTEKHistory(): List<TemporaryExposureKey> {
+        return if (enfVersion.isAtLeast(ENFVersion.V1_8)) {
+            getPreAuthTEKHistoryOnV18()
+        } else {
+            getTEKHistoryOnPreV18()
+        }
+    }
+
+    override suspend fun preAuthorizeExposureKeyHistory(): Boolean {
+        // Pre-Auth isn't available exist early
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/tekhistory/DefaultTEKHistoryProvider.kt:60]: Fix typo?

**Comment by harambasicluka** [comment]
Did some on device testing:

**Xiaomi Mi 9: Android 9; ENF 18210214000**

* Happy path:
https://user-images.githubusercontent.com/64483219/107773678-4a7c6b80-6d3e-11eb-9ef8-e7bf20df15c9.mp4

* Canceled consent, came back:
https://user-images.githubusercontent.com/64483219/107773714-59631e00-6d3e-11eb-8871-aa7414f44ac2.mp4

* Consent still valid after reactivating the in EOL:
https://user-images.githubusercontent.com/64483219/107773758-6849d080-6d3e-11eb-9184-5e313f24ad25.mp4

* If the new consent is denied the old one shows up:
https://user-images.githubusercontent.com/64483219/107773800-77308300-6d3e-11eb-9c73-a489ac20f7d7.mp4

**Pixel 3: Android 11; ENF 18205016000**

* Old ENF version didn't try to use the new consent dialog, the old one is still there:

https://user-images.githubusercontent.com/64483219/107773998-beb70f00-6d3e-11eb-83c4-c91794a976f5.mp4

**Comment by ralfgehrer** [comment]
On a Pixel (Android 10) with Play service `18210214000`
I tested all edge cases with QR codes. Even scanning a positive code, resetting the app, scanning a positive test, ... worked as expected.

Afterward, I used the Tele Tan, reset the app, and was about to use the QR Code: before scanning the code, I was not prompted to share keys BEFORE scanning the code - as I would have expected. Is this a bug or a wrong assumption of mine?

**Comment by mtwalli** [comment]
@ralfgehrer , Thanks, could you please record a video of the scenario that you think it is problematic? then I can investigate it properly

**Comment by ralfgehrer** [comment]
> @ralfgehrer , Thanks, could you please record a video of the scenario that you think it is problematic? then I can investigate it properly

I tried to reproduce the bug again and record a video, but was not able to. Maybe I did not pay super close attention earlier in the last run. But on the bright side, all seems to work very nicely. Ran about 10 more tests ;)

**Review by ralfgehrer — DISMISSED** [review]

**Review by ralfgehrer — DISMISSED** [review]

**Review by ralfgehrer — APPROVED** [review]
Thanks for the extra set of tests!

**Review by jurajkusnier — APPROVED** [review]
LGTM

**Comment by mar-v-in** [comment]
According to the first message in the PR, this feature requires *Google play services version* 18,210,214,000
The latest Google play services version according to [apkmirror.com](https://www.apkmirror.com/apk/google-inc/google-play-services/) is 210,214,000 released in Februar 2021.

So I guess you're not referring to Google play services version but something else?

**Comment by mtwalli** [comment]
> According to the first message in the PR, this feature requires _Google play services version_ 18,210,214,000
> The latest Google play services version according to [apkmirror.com](https://www.apkmirror.com/apk/google-inc/google-play-services/) is 210,214,000 released in Februar 2021.
> 
> So I guess you're not referring to Google play services version but something else?

To be precise : Exposure Notifications Version, You can check that by going to:
`Device Settings -> Google Settings -> COVID-19 Exposure Notifications -> Scroll to the bottom`
